### PR TITLE
Metadata schema guidelines for helm charts

### PR DESCRIPTION
**Pull request by hisarbalik**
kyma-project/kyma#4621

### CODE REVIEW

File: guidelines/coding-standards/helm.md [hunk 1]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
 1. Backup all existing implementations. 
 2. Remove the old CRD.
 3. Run the upgrade.
-4. Restore all CRD implementations. 
\ No newline at end of file
+4. Restore all CRD implementations. 
+
+## Defining metadata schema for Kyma charts
+
+This section cover minimal requirements and conventions of metadata schema definition for Kyma charts.
```
Comment (clebs) [inline guidelines/coding-standards/helm.md]:
```suggestion
This section covers the minimal requirements and conventions of metadata schema definition for Kyma charts. 
```

File: guidelines/coding-standards/helm.md [hunk 2]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
 3. Run the upgrade.
-4. Restore all CRD implementations. 
\ No newline at end of file
+4. Restore all CRD implementations. 
+
+## Defining metadata schema for Kyma charts
+
+This section cover minimal requirements and conventions of metadata schema definition for Kyma charts. 
+
+Schema should define important configuration parameters (not all parameters listed in `values.yaml` files) from customer perspective and which can vary depending on installation.
```
Comment (clebs) [inline guidelines/coding-standards/helm.md]:
```suggestion
The schema should define important configuration parameters (not all parameters listed in `values.yaml` files) from a customer perspective and which can vary depending on installation.
```

File: guidelines/coding-standards/helm.md [hunk 3]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+1. Each chart should have a metadata file with name `values.schema.json` placed where charts `values.yaml` (same for sub-charts) file placed. E.g. see screenshot below.
```
Comment (clebs) [inline guidelines/coding-standards/helm.md]:
```suggestion
1. Each chart should have a metadata file with name `values.schema.json` placed where the chart's `values.yaml` (same for sub-charts) file is placed. E.g. see screenshot below.
```

File: guidelines/coding-standards/helm.md [hunk 4]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
 3. Run the upgrade.
-4. Restore all CRD implementations. 
\ No newline at end of file
+4. Restore all CRD implementations. 
+
+## Defining metadata schema for Kyma charts
+
+This section covers the minimal requirements and conventions of metadata schema definition for Kyma charts. 
+
+The schema should define important configuration parameters (not all parameters listed in `values.yaml` files) from a customer perspective and which can vary depending on installation.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
The schema defines configuration parameters important from a customer perspective, as opposed to all parameters you can find in the `values.yaml` files. These parameters can vary depending on installation.
```

File: guidelines/coding-standards/helm.md [hunk 5]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+1. Each chart should have a metadata file with name `values.schema.json` placed where the chart's `values.yaml` (same for sub-charts) file is placed. E.g. see screenshot below.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
1. Make sure to place the `values.schema.json` metadata file where the chart's `values.yaml` file resides. See the screenshot for reference.
```

File: guidelines/coding-standards/helm.md [hunk 6]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. Each schema should define `desription` for detailed information about schema and helm chart.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
2. For each schema, define a description including detailed information about schema and Helm chart. 
```

File: guidelines/coding-standards/helm.md [hunk 7]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+The schema should define important configuration parameters (not all parameters listed in `values.yaml` files) from a customer perspective and which can vary depending on installation.
+
+1. Each chart should have a metadata file with name `values.schema.json` placed where the chart's `values.yaml` (same for sub-charts) file is placed. E.g. see screenshot below.
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. Each schema should define `desription` for detailed information about schema and helm chart. 
+
+3. Schema definition does not support dot `(.)` notation which mean if you have nested value properties schema definition should define object structure.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
3. Schema definition does not support dot `(.)` notation. This means that if you have nested value properties, your schema definition must define the object structure.
```

File: guidelines/coding-standards/helm.md [hunk 8]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+E.g.  helm chart value `.Values.loki.port` should be modeled in schema example below.
+
+4. Each schema `object` definition should define at least a `description` to explain configuration section
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
4. For each schema `object`, define a **description**  to explain the purpose of the schema definition.
```

File: guidelines/coding-standards/helm.md [hunk 9]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+5. Each configuration property should declare:
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
5. For each configuration property, declare:
```

File: guidelines/coding-standards/helm.md [hunk 10]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  - A `description` to explain purpose of configuration property
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  - A **description** to explain the configuration purpose.
```

File: guidelines/coding-standards/helm.md [hunk 11]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  - A `default` to provide default value of configuration property
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  - A **default** value.
```

File: guidelines/coding-standards/helm.md [hunk 12]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  - A `type` to declare data type of configuration property value
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  - A data **type**.
```

File: guidelines/coding-standards/helm.md [hunk 13]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  - A `examples` if required, a list of possible example values, e.g. storage types supported (see example schema below)
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  -  An **examples** property to list of possible example values, such as supported storage types. This property is optional.
```

File: guidelines/coding-standards/helm.md [hunk 14]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+Example chart values.yaml file
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
An example chart `values.yaml` file looks as follows:

```

File: guidelines/coding-standards/helm.md [hunk 15]
```diff
@@ -58,4 +58,103 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+The schema should define important configuration parameters (not all parameters listed in `values.yaml` files) from a customer perspective and which can vary depending on installation.
+
+1. Each chart should have a metadata file with name `values.schema.json` placed where the chart's `values.yaml` (same for sub-charts) file is placed. E.g. see screenshot below.
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. Each schema should define `desription` for detailed information about schema and helm chart. 
+
+3. Schema definition does not support dot `(.)` notation which mean if you have nested value properties schema definition should define object structure.
+E.g.  helm chart value `.Values.loki.port` should be modeled in schema example below.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
See the **.Values.loki.port** property for a sample object structure defined within a schema. 
```

File: guidelines/coding-standards/helm.md [hunk 16]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+1. Make sure to place the `values.schema.json` metadata file where the chart's `values.yaml` file resides. See the screenshot for reference.
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. For each schema, define a description including detailed information about schema and Helm chart. 
+
+3. Schema definition does not support dot `(.)` notation. This means that if you have nested value properties, your schema definition must define the object structure.
+See the **.Values.loki.port** property for a sample object structure defined within a schema. 
+
+4. For each schema `object`, define a **description**  to explain the purpose of the schema definition.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
4. For each schema object, define a **description**  to explain its purpose.
```

File: guidelines/coding-standards/helm.md [hunk 17]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+5. For each configuration property, declare:
+  - A **description** to explain the configuration purpose.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  - A **description** property to explain the configuration purpose.
```

File: guidelines/coding-standards/helm.md [hunk 18]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  - A **default** value.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  - A **default** property and its value.
```

File: guidelines/coding-standards/helm.md [hunk 19]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  - A data **type**.
+  -  An **examples** property to list of possible example values, such as supported storage types. This property is optional.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
  - An **examples** property to list of possible example values, such as supported storage types. This property is optional.
```

File: guidelines/coding-standards/helm.md [hunk 20]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+The schema defines configuration parameters important from a customer perspective, as opposed to all parameters you can find in the `values.yaml` files. These parameters can vary depending on installation.
+
+1. Make sure to place the `values.schema.json` metadata file where the chart's `values.yaml` file resides. See the screenshot for reference.
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. For each schema, define a description including detailed information about schema and Helm chart. 
+
+3. Schema definition does not support dot `(.)` notation. This means that if you have nested value properties, your schema definition must define the object structure.
+See the **.Values.loki.port** property for a sample object structure defined within a schema.
```
[the remark this answers is not in the document]
Comment (hisarbalik) [reply guidelines/coding-standards/helm.md]: it's always start with .Values....

[the remark this answers is not in the document]
Comment (bszwarc) [reply guidelines/coding-standards/helm.md]: Got it 👍

File: guidelines/coding-standards/helm.md [hunk 21]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  config:
+    auth_enabled: false
+    store: inmemory
+
+promtail:
+  port: 3101
+  nameOverride: promtail
+```
+
+Example schema definition for values.yaml
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
An example `values.yaml` file looks as follows:
```

File: guidelines/coding-standards/helm.md [hunk 22]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
\ No newline at end of file
+4. Restore all CRD implementations. 
+
+## Defining metadata schema for Kyma charts
+
+This section covers the minimal requirements and conventions of metadata schema definition for Kyma charts. 
+
+The schema defines configuration parameters important from a customer perspective, as opposed to all parameters you can find in the `values.yaml` files. These parameters can vary depending on installation.
+
+1. Make sure to place the `values.schema.json` metadata file where the chart's `values.yaml` file resides. See the screenshot for reference.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
* Make sure to place the `values.schema.json` metadata file where the chart's `values.yaml` file resides. See the screenshot for reference.
```

File: guidelines/coding-standards/helm.md [hunk 23]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. For each schema, define a description including detailed information about schema and Helm chart.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
* For each schema, define a description including detailed information about schema and Helm chart. 
```

File: guidelines/coding-standards/helm.md [hunk 24]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+The schema defines configuration parameters important from a customer perspective, as opposed to all parameters you can find in the `values.yaml` files. These parameters can vary depending on installation.
+
+1. Make sure to place the `values.schema.json` metadata file where the chart's `values.yaml` file resides. See the screenshot for reference.
+
+   ![Example 1](../../assets/metadata-schema.png)
+
+2. For each schema, define a description including detailed information about schema and Helm chart. 
+
+3. Schema definition does not support dot `(.)` notation. This means that if you have nested value properties, your schema definition must define the object structure.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
* Schema definition does not support dot `(.)` notation. This means that if you have nested value properties, your schema definition must define the object structure.
```

File: guidelines/coding-standards/helm.md [hunk 25]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+See the **.Values.loki.port** property for a sample object structure defined within a schema. 
+
+4. For each schema object, define a **description**  to explain its purpose.
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
* For each schema object, define a **description**  to explain its purpose.
```

File: guidelines/coding-standards/helm.md [hunk 26]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+
+5. For each configuration property, declare:
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
* For each configuration property, declare:
```

File: guidelines/coding-standards/helm.md [hunk 27]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
-4. Restore all CRD implementations. 
\ No newline at end of file
+4. Restore all CRD implementations. 
+
+## Defining metadata schema for Kyma charts
+
+This section covers the minimal requirements and conventions of metadata schema definition for Kyma charts. 
+
+The schema defines configuration parameters important from a customer perspective, as opposed to all parameters you can find in the `values.yaml` files. These parameters can vary depending on installation.
+
```
Comment (bszwarc) [inline guidelines/coding-standards/helm.md]:
```suggestion
When defining a schema for a Kyma chart, follow these guidelines: 
```

File: guidelines/coding-standards/helm.md [hunk 28]
```diff
@@ -58,4 +58,104 @@ When a CRD is deleted, all of the associated implementations are removed, which
...
+  config:
+    auth_enabled: false
+    store: inmemory
+
+promtail:
+  port: 3101
+  nameOverride: promtail
+```
+
+An example `values.yaml` file looks as follows:
```
Comment (a-thaler) [inline guidelines/coding-standards/helm.md]: It should be "values.schema.json"

**Pull request merged**